internal/worker_versioning_rules.go

```diff
@@ -0,0 +1,338 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"errors"
+	"time"
+
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	// VersioningRamp is an interface for the different strategies of gradual workflow deployments.
+	VersioningRamp interface {
+		isValidVersioningRamp() bool
+	}
+
+	// VersioningRampByPercentage sends a proportion of the traffic to the target Build ID.
+	VersioningRampByPercentage struct {
+		// Percentage of traffic with a value in [0,100)
+		Percentage float32
+	}
+
+	// VersioningAssignmentRule is a BuildID assigment rule for a task queue.
+	// Assignment rules only affect new workflows.
+	VersioningAssignmentRule struct {
+		// The BuildID of new workflows affected by this rule.
+		TargetBuildId string
+		// A strategy for gradual workflow deployment.
+		Ramp VersioningRamp
+		// An optional registration time set by the server.
+		Timestamp *time.Time
+	}
+
+	// VersioningAssignmentRule is a BuildID redirect rule for a task queue.
+	// It changes the behavior of currently running workflows and new ones.
+	VersioningRedirectRule struct {
+		SourceBuildId string
+		TargetBuildId string
+		// An optional registration time set by the server.
+		Timestamp *time.Time
+	}
+
+	//VersioningConflictToken is a conflict token to serialize updates.
+	// An update with an old token fails with `serviceerror.FailedPrecondition`.
+	// The current token can be obtained with [GetWorkerVersioningRules], or returned by a successful [UpdateWorkerVersioningRules].
+	VersioningConflictToken struct {
+		token []byte
```

Comment:
Users cannot read this value and therefore cannot pass it around other systems if they need to. We should just give them a byte array directly, we don't need to wrap it in a struct or assume it needs to be hidden from them. This is similar to task tokens we use in other situations.

Reply:
I wanted to avoid that they can change the byte array outside the package, and we could ended up adding other things, like a timestamp to expire the token

Reply:
I don't think we need to avoid that. No need to hide this (and it may have value for a user to store the byte array if they want). This same thing applies to "task token".

Reply:
I prefer to wrap it, maybe my paranoid years of working on security, it also hides the implementation of the conflict token, and makes it easier to extend it.

Reply:
Please let users have access to this token to do what they want with it